### DISCUSSION TITLE
PnP Core SDK integration with Site.Selected Graph Permission

### DISCUSSION THREAD

**Discussion by somdeepgniyogi** [pnp/pnpcore#709] [Q&A]
Hi,

We are working on a project which is an Azure App connecting to Sharepoint Online.
The permissions allowed by the Organization is only Site.Selected Application permission.

The above permission setup does not work with PnP.Core context.

Is there any update about the future availability of this integration.?

BR

**Comment by jansenbe** [comment]
@somdeepgniyogi : currently Sites.Selected does only work with Graph API calls and not with SPO REST or CSOM calls. Given that PnP Core SDK uses REST, Graph and CSOM it currently does not function with Sites.Selected. We're however updating Sites.Selected to also work with SPO REST and CSOM calls and at that point things should just work.

**Reply by jansenbe** [reply]
Moved this from issues to discussions as this is not really an SDK issue

**Reply by somdeepgniyogi** [reply]
Dear @jansenbe , 
Thanks a lot for your reply. We are in planning phase for the project and considering various options. PnP Core is the option of choice for us.

Do you have any idea as of when this Sites.Selected will have the feature to work with Rest/CSOM as otherwise we have to look for other alternatives. :(

BR

**Reply by jansenbe** [reply]
I'll cross check with the team on timelines @somdeepgniyogi , in the meanwhile I would advice to just use tenant scoped application permissions. If your application can fulfill all it's needs with Graph Only using the Graph SDK will today work with Sites.Selected, but from my experience (backend) applications quite often need more than what Graph currently offers, hence REST/CSOM are needed.

**Reply by somdeepgniyogi** [reply]
Dear @jansenbe,
 I would have been very happy if my organization would have allowed Tenant Level  Application permissions .
Unfortunately the security team is not in favor of it. The application have mixed permissions where Tenant Level Delegated permissions are allowed. However there are some scenarios which need elevation based on Application Permission and only for those we have to use Sites.Selected.

BR

**Reply by jansenbe** [reply]
@somdeepgniyogi : sites.selected support for REST/CSOM calls rolled out during year end, so it should be available. Are you testing in a regular tenant or in special one (US Gov, DoD)? If you're testing in a regular tenant then please share more details on the error you're getting

**Reply by jansenbe** [reply]
@somdeepgniyogi : also, when you test don't forget to use sites.selected for both the Graph as SharePoint permissions:

![image](https://user-images.githubusercontent.com/7451219/148812658-33a4e1f5-eeb1-4358-9827-e48bf9d56aea.png)

**Reply by somdeepgniyogi** [reply]
Dear @jansenbe ,

I am using the following code

 public async Task<PnPContext> CreateAppSiteContext()
        {
            var siteUrl = new Uri(_pnpCoreOptions.Sites["Site"].SiteUrl);       
            string[] scopes = new string[] { $"https://graph.microsoft.com/.default" };           
            try
            {            
              
                return await _pnpContextFactory.CreateAsync(siteUrl,
                               new ExternalAuthenticationProvider((resourceUri, scopes) =>
                               {
                                    return _tokenAcquisition.GetAccessTokenForAppAsync(scopes.First()).GetAwaiter().GetResult();                                   
                               }
                       )).ConfigureAwait(false);
              
            }
            catch (Exception ex)
            {
                return null;
            }
        }

**Exception : SharePoint Rest service exception**

Can you kindly tell me what is wrong?

BR

**Reply by jansenbe** [reply]
You're getting your access token via something "external" and seems you're only requesting an access token for Graph. You would also need to request an access token for REST and have the proper setup done in the Azure AD app to support that (see my previous screenshot).

**Reply by somdeepgniyogi** [reply]
Dear @jansenbe ,

I have added the required permission Sites.Selected for both Graph and Sharepoint in Azure  App registration. Can you kindly guide me what am I missing. I followed the screenshot you provided.

BR

**Reply by jansenbe** [reply]
Please check if the returned access token is having the correct scopes?

**Reply by somdeepgniyogi** [reply]
Dear @jansenbe ,

I have attaching the token scopes below.

<img width="664" alt="Untitled" src="https://user-images.githubusercontent.com/6120892/149154555-777861cf-4135-4285-8354-555c599338f4.png">



BR

**Reply by jansenbe** [reply]
@somdeepgniyogi : In your sample code at this line `return _tokenAcquisition.GetAccessTokenForAppAsync(scopes.First()).GetAwaiter().GetResult();                                  ` do you see the right scopes being requested - you should see `https://graph.microsoft.com/.default` whenever PnP Core SDK is making a graph call or `https://yourtenant.sharepoint.com/.default` for regular tenants. So you token acquisition code (which is external to PnP Core SDK as you're using the `ExternalAuthenticationProvider`) must be able to return tokens for both scopes.

**Reply by somdeepgniyogi** [reply]
Dear @jansenbe ,

Tried with both scopes below

string[] scopes = new string[] { $"https://graph.microsoft.com/.default" };
string[] scopes1= new string[] { $"https://xxx.sharepoint.com/.default" }; 

Exception : SharePoint Rest service exception

BR

**Reply by jansenbe** [reply]
@somdeepgniyogi : and your sample works with regular permissions? I assume you did grant the site you're using access via sites.selected? See https://www.leonarmston.com/2021/03/testing-out-the-new-microsoft-graph-sharepoint-specific-site-collection-app-permissions-with-pnp-powershell/ and https://ashiqf.com/2021/03/15/how-to-use-microsoft-graph-sharepoint-sites-selected-application-permission-in-a-azure-ad-application-for-more-granular-control/ for instructions on to setup that.

You can also do a fiddler trace to see what goes on across the wire, it will show the made requests with their access tokens

**Reply by somdeepgniyogi** [reply]
Dear @jansenbe ,

The exception comes as below

HttpResponseCode: 401
Code: 
Message: Unsupported app only token.
ClientRequestId: f19e16a0-8075-3000-a1be-01d5bdc0d8a9

Can you kindly help me here.

I have set up the sites.selected permission following the links you provided above.

BR,
Somdeep

**Reply by jansenbe** [reply]
@somdeepgniyogi : are you using a regular tenant?

**Reply by somdeepgniyogi** [reply]
Dear @jansenbe ,

Yes I am using a regular tenant.

BR,
Somdeep

**Reply by jansenbe** [reply]
@somdeepgniyogi : Are you using a certificate for authentication? App-Only for SharePoint APIs requires certificate based auth, using client+secret will throw an exception like you're seeing.

**Reply by somdeepgniyogi** [reply]
Deat @jansenbe ,

No I am not using certificate. I'll try with MSI.

Thanks a lot for you help.

BR

**Reply by jansenbe** [reply]
@somdeepgniyogi : does it work now?

**Reply by jansenbe** [reply]
@somdeepgniyogi : if you got things working, please mark this one as answered. If not, let's further investigate what's needed.

**Comment by krymancer** [comment]
> @somdeepgniyogi : Are you using a certificate for authentication? App-Only for SharePoint APIs requires certificate based auth, using client+secret will throw an exception like you're seeing.

How to use client+secret? I tried to replicate some of the samples and can't get to work
